woocommerce/google-listings-and-ads
# Add Normalizer Polyfill

### Changes proposed in this Pull Request:



When I was trying to sync products I was getting the following error:

```
The "normalizer" option must be a valid callable ("string" given). in /var/www/html/wp-content/plugins/google-listings-and-ads/vendor/symfony/validator/Constraints/Url.php:54 Stack trace: #0 /var/www/html/wp-content/plugins/google-listings-and-ads/src/Product/WCProductAdapter.php(756):
```

The error was caused because the [Int](https://php.net/intl) extension was not installed hence the `Normalizer` class was not found.  

This PR adds the Normalizer polyfill, based off https://github.com/symfony/polyfill-intl-normalizer, unfortunately we can't add the package as a dependency using `composer require symfony/polyfill-intl-normalizer` as it causes an error when using with woorelease. See more context: https://github.com/woocommerce/google-listings-and-ads/issues/331#issuecomment-838372148 and https://github.com/woocommerce/google-listings-and-ads/pull/579.

Because of this issue, the polyfill needs to be added manually to our repository. I added only the methods needed to run the plugin.

### Screenshots:

https://user-images.githubusercontent.com/2488994/178124584-29ceadfa-1f42-4d8d-ac86-c069f190fd77.mp4


### Detailed test instructions:


1. Disable the extension `intl` extension by adding a semicolon at the begging, for example: `;extension=intl.so`. In my case I disabled it going to the following file: `/usr/local/etc/php/conf.d/docker-php-ext-intl.ini`
2. Go to `/wp-admin/admin.php?page=connection-test-admin-page`
3. Sync a product, without clicking `async`
4. The following error should be displayed: `The "normalizer" option must be a valid callable.....`
5. Check out this PR
6. run `composer dump-autoload` to create the autoload files.
7. Repeat steps from 2 to 4
8. No errors should be displayed

### Additional details:

- I applied phpcs in the new files but I observed that adding `Resources/unidata` files was increasing considerably the time to run phpcs. In my local setup went from 15s to more than 1min therefore I set `// phpcs:ignoreFile` in those files.
- I applied phpcs `Normalizer.php` in the places where it was safe to do it, I did not want it to replace all variables to be snake_case as I think it could increase the risk to create an unnecessary bug. 
- I added a new polyfill `mb_internal_encoding` to the class MBString as they are used in the `Normalizer` class.

### Changelog entry

>Add - Normalizer Polyfill
